File: ascii_fields/animations/night_sky.py
import math

from ..core import Animation, clamp, render_field, star_noise
from ..noise import fbm


class NightSkyAnimation(Animation):
  """A still night sky: twinkling stars, a soft Milky Way band, and the odd
  shooting star streaking across."""

  thresholds = [
    (0.06, " "), (0.16, "."), (0.30, ":"), (0.45, "-"), (0.60, "+"),
    (0.74, "*"), (0.86, "o"), (0.94, "#"), (1.01, "@"),
  ]

  def render(self, width, height, elapsed, phase, options):
    density = max(0.4, options.scale)
    # Milky-way band: a diagonal swathe of faint fbm haze.
    band_angle = 0.5
    ca, sa = math.cos(band_angle), math.sin(band_angle)
    meteor = self._meteor(elapsed, width, height)
    grid = []
    for row in range(height):
      ny = (row / max(1, height - 1) - 0.5) * 2.0
      line = []
      for col in range(width):
        nx = (col / max(1, width - 1) - 0.5) * 2.0
        # distance from the band centre line, rotated
        across = nx * sa - ny * ca
        haze = math.exp(-(across * across) / 0.12)
        milky = haze * (0.10 + 0.26 * fbm(nx * 3.0 + 4.0, ny * 3.0, octaves=4, seed=7))
        level = 0.04 + milky
        # Stars: sparse, twinkling.
        n = star_noise(col, row)
        if n > 0.991:
          mag = 0.55 + 0.45 * star_noise(col + 3, row + 11)
          twinkle = 0.6 + 0.4 * math.sin(elapsed * (1.5 + 4.0 * n) + col * 0.7 + row)
          level = max(level, mag * twinkle * (0.7 + 0.3 * density))
        elif n > 0.975:
          level = max(level, 0.22 + 0.10 * math.sin(elapsed * 2.0 + col))
        if meteor is not None:
          level = max(level, self._meteor_brightness(meteor, col, row))
        line.append(clamp(level))
      grid.append(line)
    return render_field(width, height, grid, options, self)

  def _meteor(self, elapsed, width, height):
    period = 6.5
    idx = int(elapsed / period)
    local = elapsed - idx * period
    duration = 1.1
    if local > duration:
      return None
    r = star_noise(idx * 17 + 3, idx * 5 + 1)
    r2 = star_noise(idx * 9 + 7, idx * 13 + 2)
    start_x = r * width
    start_y = r2 * height * 0.5
    dx = (0.7 + 0.5 * r2) * width
    dy = (0.4 + 0.4 * r) * height
    p = local / duration
    head_x = start_x + dx * p
    head_y = start_y + dy * p
    return (head_x, head_y, dx, dy, 1.0 - p)

  def _meteor_brightness(self, meteor, col, row):
    head_x, head_y, dx, dy, fade = meteor
    length = math.hypot(dx, dy)
    ux, uy = dx / length, dy / length
    rx, ry = col - head_x, row - head_y
    along = rx * ux + ry * uy            # negative = behind head (the tail)
    perp = abs(rx * uy - ry * ux)
    if along > 0.6 or along < -10.0 or perp > 1.4:
      return 0.0
    tail = math.exp(along * 0.45) if along < 0 else 1.0
    core = math.exp(-(perp * perp) / 0.5)
    return clamp(tail * core * fade * 1.2)

File: src/animations/night_sky.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use crate::noise::{fbm, star_noise};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ice" };
const TH: &[(f64, char)] = &[
  (0.06, ' '),
  (0.16, '.'),
  (0.30, ':'),
  (0.45, '-'),
  (0.60, '+'),
  (0.74, '*'),
  (0.86, 'o'),
  (0.94, '#'),
  (1.01, '@'),
];

pub struct NightSky;

impl Animation for NightSky {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width;
    let h = ctx.height;
    let density = ctx.options.scale.max(0.4);
    let mut grid = vec![0.0_f64; w * h];
    let band_angle = 0.5_f64;
    let ca = band_angle.cos();
    let sa = band_angle.sin();
    let meteor = meteor_state(ctx.elapsed, w, h);
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    for row in 0..h {
      let ny = (row as f64 / dh - 0.5) * 2.0;
      let base = row * w;
      for col in 0..w {
        let nx = (col as f64 / dw - 0.5) * 2.0;
        let across = nx * sa - ny * ca;
        let haze = (-(across * across) / 0.12).exp();
        let milky = haze * (0.10 + 0.26 * fbm(nx * 3.0 + 4.0, ny * 3.0, 4));
        let mut level = 0.04 + milky;
        let n = star_noise(col as i64, row as i64);
        if n > 0.991 {
          let mag = 0.55 + 0.45 * star_noise(col as i64 + 3, row as i64 + 11);
          let twinkle = 0.6 + 0.4 * (ctx.elapsed * (1.5 + 4.0 * n) + col as f64 * 0.7 + row as f64).sin();
          level = level.max(mag * twinkle * (0.7 + 0.3 * density));
        } else if n > 0.975 {
          level = level.max(0.22 + 0.10 * (ctx.elapsed * 2.0 + col as f64).sin());
        }
        if let Some(m) = &meteor {
          let mb = meteor_brightness(m, col as f64, row as f64);
          if mb > level {
            level = mb;
          }
        }
        grid[base + col] = clamp(level);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}

struct Meteor {
  head_x: f64,
  head_y: f64,
  dx: f64,
  dy: f64,
  fade: f64,
}

fn meteor_state(elapsed: f64, w: usize, h: usize) -> Option<Meteor> {
  let period = 6.5;
  let duration = 1.1;
  let idx = (elapsed / period) as i64;
  let local = elapsed - idx as f64 * period;
  if local > duration {
    return None;
  }
  let r = star_noise(idx * 17 + 3, idx * 5 + 1);
  let r2 = star_noise(idx * 9 + 7, idx * 13 + 2);
  let start_x = r * w as f64;
  let start_y = r2 * h as f64 * 0.5;
  let dx = (0.7 + 0.5 * r2) * w as f64;
  let dy = (0.4 + 0.4 * r) * h as f64;
  let p = local / duration;
  Some(Meteor { head_x: start_x + dx * p, head_y: start_y + dy * p, dx, dy, fade: 1.0 - p })
}

// `along > 0.6 || along < -10.0` mirrors "outside the meteor's body along
// the trail direction"; converting to !RangeInclusive::contains hides intent.
#[allow(clippy::manual_range_contains)]
fn meteor_brightness(m: &Meteor, col: f64, row: f64) -> f64 {
  let length = (m.dx * m.dx + m.dy * m.dy).sqrt();
  let ux = m.dx / length;
  let uy = m.dy / length;
  let rx = col - m.head_x;
  let ry = row - m.head_y;
  let along = rx * ux + ry * uy;
  let perp = (rx * uy - ry * ux).abs();
  if along > 0.6 || along < -10.0 || perp > 1.4 {
    return 0.0;
  }
  let tail = if along < 0.0 { (along * 0.45).exp() } else { 1.0 };
  let core = (-(perp * perp) / 0.5).exp();
  (tail * core * m.fade * 1.2).clamp(0.0, 1.0)
}
